### PR TITLE
{chem}[foss/2019b,fosscuda/2019b] NAMD v2.14

### PR DESCRIPTION
**Pull request by edmondac**
This requires ~~https://github.com/easybuilders/easybuild-easyblocks/pull/2113~~

(created using `eb --new-pr`)

edit: now also requires ~~https://github.com/easybuilders/easybuild-easyblocks/pull/2118~~

### CODE REVIEW

**Comment by edmondac** [comment]
Test report by @bear-rsg
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
bear-pg0211u03a.bear.cluster - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, Python 3.6.8
See https://gist.github.com/06a5a2bd7adf86cd3c12d59c45adaab1 for a full test report.

File: easybuild/easyconfigs/n/NAMD/NAMD-2.14-foss-2019b-mpi.eb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    '34044d85d9b4ae61650ccdba5cda4794088c3a9075932392dd0752ef8c049235',  # NAMD_2.14_Source.tar.gz
+    '6644b13560829e22ac5e5ceedde73b52928ef312a1f3f3775020f258e8150a4e',  # charm_6.10.2.patch
+    '0705e8454bec1fcfd20f65e12d9b9ebc70080fcd2efba9053e3931254d799a28',  # NAMD-2.14_Linux-POWER-cuda.patch
+]
+
+# /bin/csh is required by 'config' script
+osdependencies = ['tcsh']
+
+# Hard to make charm build with gcc on POWER, so we don't currently try.
+charm_arch = "mpi-linux-x64_64"
```
Comment (edmondac) [inline easybuild/easyconfigs/n/NAMD/NAMD-2.14-foss-2019b-mpi.eb:32]:
```suggestion
charm_arch = "mpi-linux-x86_64"
```

**Comment by edmondac** [comment]
Test report by @bear-rsg
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bear-pg0305u03a.bear.cluster - Linux RHEL 7.6, POWER, 8335-GTX, Python 3.6.8
See https://gist.github.com/d4de4a2bd99dcca484415b2e8ba5237d for a full test report.

**Comment by edmondac** [comment]
Test report by @bear-rsg
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bear-pg0211u03a.bear.cluster - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, Python 3.6.8
See https://gist.github.com/80e8f57a62a266dfca5f2c178124071c for a full test report.

**Comment by edmondac** [comment]
Test report by @bear-rsg
**SUCCESS**
Build succeeded for 5 out of 5 (2 easyconfigs in this PR)
bear-pg0212u27a.bear.cluster - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz, Python 3.6.8
See https://gist.github.com/2455105e83e5b28fbba886eccb452bd0 for a full test report.

File: easybuild/easyconfigs/n/NAMD/NAMD-2.14_Linux-POWER-cuda.patch [hunk 2]
```diff
@@ -0,0 +1,22 @@
+--- arch/Linux-POWER.cuda.orig	2020-08-05 18:51:20.000000000 +0100
```
Comment (smoors) [inline easybuild/easyconfigs/n/NAMD/NAMD-2.14_Linux-POWER-cuda.patch:1]: @edmondac please add author and short description to the top of the patches?

Comment (edmondac) [reply easybuild/easyconfigs/n/NAMD/NAMD-2.14_Linux-POWER-cuda.patch:5]: Done

File: easybuild/easyconfigs/n/NAMD/NAMD-2.14-foss-2019b-mpi.eb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+]
+
+dependencies = [
+    ('Tcl', '8.6.9'),
+    ('FFTW', '3.3.8'),
+]
+checksums = [
+    '34044d85d9b4ae61650ccdba5cda4794088c3a9075932392dd0752ef8c049235',  # NAMD_2.14_Source.tar.gz
+    '6644b13560829e22ac5e5ceedde73b52928ef312a1f3f3775020f258e8150a4e',  # charm_6.10.2.patch
+    '0705e8454bec1fcfd20f65e12d9b9ebc70080fcd2efba9053e3931254d799a28',  # NAMD-2.14_Linux-POWER-cuda.patch
```
Comment (edmondac) [inline easybuild/easyconfigs/n/NAMD/NAMD-2.14-foss-2019b-mpi.eb:25]:
```suggestion
    'db4aeb482dfa805c859ea18940026395763169e0257401ee5341ca550029031c',  # NAMD-2.14_Linux-POWER-cuda.patch
```

File: easybuild/easyconfigs/n/NAMD/NAMD-2.14-fosscuda-2019b.eb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+sources = [{'filename': 'NAMD_%(version)s_Source.tar.gz'}]
+patches = [
+    "charm_6.10.2.patch",  # fix from latest charm github
+    "%(name)s-%(version)s_Linux-POWER-cuda.patch",  # support for GCC 8+ on POWER
+]
+checksums = [
+    '34044d85d9b4ae61650ccdba5cda4794088c3a9075932392dd0752ef8c049235',  # NAMD_2.14_Source.tar.gz
+    '6644b13560829e22ac5e5ceedde73b52928ef312a1f3f3775020f258e8150a4e',  # charm_6.10.2.patch
+    '0705e8454bec1fcfd20f65e12d9b9ebc70080fcd2efba9053e3931254d799a28',  # NAMD-2.14_Linux-POWER-cuda.patch
```
Comment (edmondac) [inline easybuild/easyconfigs/n/NAMD/NAMD-2.14-fosscuda-2019b.eb:19]:
```suggestion
    'db4aeb482dfa805c859ea18940026395763169e0257401ee5341ca550029031c',  # NAMD-2.14_Linux-POWER-cuda.patch
```

**Comment by edmondac** [comment]
Test report by @bear-rsg
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bear-pg0305u05a.bear.cluster - Linux RHEL 7.6, 8335-GTX, Python 2.7.5
See https://gist.github.com/bc4ef9662180b2924d732ccc182bb2f0 for a full test report.

File: easybuild/easyconfigs/n/NAMD/NAMD-2.14_Linux-POWER-cuda.patch [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+--- arch/Linux-POWER.cuda.orig	2020-08-05 18:51:20.000000000 +0100
++++ arch/Linux-POWER.cuda	2020-08-11 11:24:49.471069513 +0100
+@@ -13,5 +13,6 @@
+ CUDA=$(CUDAFLAGS) -I. $(CUBINCL) $(CUDAINCL) 
+ CUDACC=$(CUDADIR)/bin/nvcc -Xcompiler "-m64"
+ 
+-CUDACCOPTS=-O3 --maxrregcount 48 $(CUDAGENCODE) $(CUDA) -use_fast_math
++# For POWER and gcc 8+ we need to set -std=c++11 to avoid 'error: identifier “__ieee128” is undefined'
++CUDACCOPTS=-O3 --maxrregcount 48 $(CUDAGENCODE) $(CUDA) -use_fast_math -std=c++11
```
Comment (boegel) [inline easybuild/easyconfigs/n/NAMD/NAMD-2.14_Linux-POWER-cuda.patch:13]: @edmondac Can't this be done via `toolchainopts`?

```python
toolchainopts = {'cstd': 'c++11'}`?
```

File: easybuild/easyconfigs/n/NAMD/charm_6.10.2.patch [hunk 6]
```diff
@@ -0,0 +1,11 @@
+--- charm-6.10.2/src/conv-core/memoryaffinity.C.orig	2020-08-11 10:32:08.024080784 +0100
++++ charm-6.10.2/src/conv-core/memoryaffinity.C	2020-08-11 10:32:28.356832962 +0100
+@@ -182,7 +182,7 @@
+         for (i=0; i<strlen((const char *)nodemap); i++) {
+             if (nodemap[i]==',') nodemapArrSize++;
+         }
+-        nodemapArr = malloc(nodemapArrSize*sizeof(int));
++        nodemapArr = (int *) malloc(nodemapArrSize*sizeof(int));
```
Comment (boegel) [inline easybuild/easyconfigs/n/NAMD/charm_6.10.2.patch:8]: If we can avoid dragging along a patch, I think we should.

I fixed this issue by passing `-fpermissive` as additional compiler option for Charm++, see #11114 (which requires https://github.com/easybuilders/easybuild-easyblocks/pull/2118)

File: easybuild/easyconfigs/n/NAMD/NAMD-2.14_Linux-POWER-cuda.patch (hunk 5 above)
Comment (edmondac) [reply easybuild/easyconfigs/n/NAMD/NAMD-2.14_Linux-POWER-cuda.patch]: Done

File: easybuild/easyconfigs/n/NAMD/charm_6.10.2.patch (hunk 6 above)
Comment (edmondac) [reply easybuild/easyconfigs/n/NAMD/charm_6.10.2.patch]: Done

**Comment by edmondac** [comment]
Test report by @bear-rsg
**FAILED**
Build succeeded for 1 out of 2 (2 easyconfigs in this PR)
bear-pg0212u27a.bear.cluster - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz, Python 3.6.8
See https://gist.github.com/82f6dd8a29341d43061a9e4131c0c224 for a full test report.

**Comment by edmondac** [comment]
Test report by @bear-rsg
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in this PR)
bear-pg0212u27a.bear.cluster - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz, Python 3.6.8
See https://gist.github.com/5be70dad03574f822ef517727753e3fc for a full test report.

**Comment by edmondac** [comment]
Test report by @bear-rsg
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
bear-pg0305u03a.bear.cluster - Linux RHEL 7.6, POWER, 8335-GTX, Python 3.6.8
See https://gist.github.com/74340fcfa6c11096d4a5b77b8efc7045 for a full test report.

**Comment by edmondac** [comment]
Test report by @bear-rsg
**FAILED**
Build succeeded for 0 out of 2 (2 easyconfigs in this PR)
bber0501u03a.bb2.cluster - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) CPU E5-2690 v3 @ 2.60GHz, Python 3.6.8
See https://gist.github.com/8f363581513af75c5390667580ed4c99 for a full test report.

**Comment by edmondac** [comment]
Test report by @bear-rsg
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bear-pg0305u03a.bear.cluster - Linux RHEL 7.6, POWER, 8335-GTX, Python 3.6.8
See https://gist.github.com/acf663b78466af7ddc0b32b49f6df187 for a full test report.

**Comment by edmondac** [comment]
Test report by @bear-rsg
**FAILED**
Build succeeded for 0 out of 2 (2 easyconfigs in this PR)
bber0501u03a.bb2.cluster - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) CPU E5-2690 v3 @ 2.60GHz, Python 3.6.8
See https://gist.github.com/692efe7c036fb5d252c6314c0aaff3a4 for a full test report.

**Comment by edmondac** [comment]
Test report by @bear-rsg
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in this PR)
bear-pg0212u27a.bear.cluster - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz, Python 3.6.8
See https://gist.github.com/39b8f456bb99a57428577505b2d0cf68 for a full test report.

**Comment by edmondac** [comment]
Test report by @bear-rsg
**FAILED**
Build succeeded for 1 out of 2 (2 easyconfigs in this PR)
bear-pg0212u27a.bear.cluster - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz, Python 3.6.8
See https://gist.github.com/09f943808f8099732cf5deb89a151204 for a full test report.

**Comment by edmondac** [comment]
Test report by @bear-rsg
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
bear-pg0212u27a.bear.cluster - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz, Python 3.6.8
See https://gist.github.com/e01b915f64d8bd074430aed88b045ae1 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 2 (2 easyconfigs in this PR)
node3309.joltik.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) Gold 6242 CPU @ 2.80GHz (cascadelake), Python 3.6.8
See https://gist.github.com/7a0d08ed127b39d02dcab8ba56404c33 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in this PR)
node3309.joltik.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) Gold 6242 CPU @ 2.80GHz (cascadelake), Python 3.6.8
See https://gist.github.com/43d95ac95dd48881e76b7a4a47e594f5 for a full test report.

**Comment by boegel** [comment]
@boegelbot please test @ generoso EB_ARGS="NAMD-2.14-foss-2019b-mpi.eb"

**Comment by boegel** [comment]
Going in, thanks @edmondac!

**Pull request merged**